tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
```

Comment:
Also, looking at `graceful_shutdown.h`:

```
void * graceful_shutdown_register(callback, caller_cookie);
void graceful_shutdown_shutdown(cookie);
```

Wouldn't it be nice to distinguish between the "graceful_shutdown cookie", and "the cookie we pass to the shutdown callback" ?